graph/src/util/bounded_queue.rs

```diff
@@ -0,0 +1,143 @@
+            queue: Mutex::new(VecDeque::with_capacity(capacity)),
+            push_semaphore: Semaphore::new(capacity),
+            pop_semaphore: Semaphore::new(0),
+        }
+    }
+
+    /// Get an item from the queue. If the queue is currently empty
+    /// this method blocks until an item is available.
+    pub async fn pop(&self) -> T {
+        let permit = self.pop_semaphore.acquire().await.unwrap();
+        let item = self
+            .queue
+            .lock()
+            .unwrap()
+            .pop_front()
+            .expect("the queue is not empty");
+        permit.forget();
+        self.push_semaphore.add_permits(1);
+        item
+    }
+
+    /// Take an item from the front of the queue and return a copy. If the
+    /// queue is currently empty this method blocks until an item is
+    /// available.
+    pub async fn peek(&self) -> T {
+        let _permit = self.pop_semaphore.acquire().await.unwrap();
+        let queue = self.queue.lock().unwrap();
+        let item = queue.front().expect("the queue is not empty");
+        item.clone()
+    }
+
+    /// Push an item into the queue. If the queue is currently full this method
+    /// blocks until an item is available
+    pub async fn push(&self, item: T) {
+        let permit = self.push_semaphore.acquire().await.unwrap();
+        self.queue.lock().unwrap().push_back(item);
+        permit.forget();
+        self.pop_semaphore.add_permits(1);
+    }
+
+    pub async fn wait_empty(&self) {
+        self.push_semaphore
+            .acquire_many(self.capacity as u32)
+            .await
+            .map(|_| ())
+            .expect("we never close the push_semaphore")
+    }
+
+    pub fn len(&self) -> usize {
+        self.queue.lock().unwrap().len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.queue.lock().unwrap().is_empty()
+    }
+
+    pub fn capacity(&self) -> usize {
+        self.capacity
+    }
+
+    /// Iterate over the entries in the queue from newest to oldest entry
+    /// atomically, applying `f` to each entry and returning the first
+    /// result that is not `None`.
+    ///
+    /// This method locks the queue while it is executing, and `f` should
+    /// therefore not do any slow work.
+    pub fn find_map<B, F>(&self, f: F) -> Option<B>
+    where
+        F: FnMut(&T) -> Option<B>,
+    {
+        let queue = self.queue.lock().unwrap();
+        queue.iter().rev().find_map(f)
+    }
+
+    /// Iterate over the entries in the queue from newest to oldest entry
+    /// atomically, applying `f` to each entry and returning the result of
+    /// the last invocation of `f`.
+    ///
+    /// This method locks the queue while it is executing, and `f` should
+    /// therefore not do any slow work.
+    pub fn fold<B, F>(&self, init: B, f: F) -> B
+    where
+        F: FnMut(B, &T) -> B,
+    {
+        let queue = self.queue.lock().unwrap();
+        queue.iter().rev().fold(init, f)
+    }
+
+    pub async fn clear(&self) {
```

Comment:
The way we use it wouldn't manifest this, but in principle calling `clear` and `pop` concurrently could cause pop to panic, since `clear` will empty the queue without first acquiring the permits. If it tried to acquire the permits first we'd have a different potential bug, where the permit count may change between it being checked and the permits actually being acquired, and then `clear` could deadlock.

I'm thinking that `BoundedQueue` shouldn't intrinsically support `clear`, but rather the writer should call `pop` in a loop until the queue is empty when clearing it. This way we can justify the assumption that neither `pop` nor `push` are being called concurrently to this.

Reply:
I added a `try_pop` (non-blocking `pop`) and made `clear` a wrapper around calling that in a loop; since there's no `await` on that code path now, there shouldn't be any possibility for races.